File: MusicDemoOC/MusicDemo/Controller/DetailViewController.m
//
//  DetailViewController.m
//  MusicDemo
//
//  Created by 沈凯 on 2017/11/8.
//  Copyright © 2017年 Ssky. All rights reserved.
//

#import "DetailViewController.h"
#import <MediaPlayer/MediaPlayer.h>
#import <AVFoundation/AVFoundation.h>
#import "CALayer+RotationAnimation.h"
#import "NSString+ConversionTime.h"
#import "NSTimer+Operation.h"
#import "LrcView.h"

@interface DetailViewController ()<AVAudioPlayerDelegate, UIScrollViewDelegate>
//歌名
@property (weak, nonatomic) IBOutlet UILabel *MusicName;
//歌者
@property (weak, nonatomic) IBOutlet UILabel *singer;
//图片
@property (weak, nonatomic) IBOutlet UIImageView *img;
//进度
@property (weak, nonatomic) IBOutlet UISlider *progress;
//播放按钮
@property (weak, nonatomic) IBOutlet UIButton *playBtn;
//当前时间
@property (weak, nonatomic) IBOutlet UILabel *currentTime;
//总体时间
@property (weak, nonatomic) IBOutlet UILabel *totalTIme;
//播放器
@property (strong, nonatomic) AVAudioPlayer *audioPlayer;
//进度的定时器
@property (strong, nonatomic) NSTimer *progressTimer;
//歌词
@property (weak, nonatomic) IBOutlet LrcLabel *lrcLabel;
//歌词视图
@property (weak, nonatomic) IBOutlet LrcView *lrcView;
//歌词定时器
@property (strong, nonatomic) CADisplayLink *lrcTimer;
@end

@implementation DetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    UI设置
    [self setUI];
//    播放音乐
    [self playMusic];
//    设置进度定时器
    [self addProgressTimer];
//    设置歌词定时器
    [self addLrcTimer];
}
- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    [self stopTimer:2];
}
#pragma mark UI设置
- (void)setUI
{
    [self setUIMessageWithModel:self.model];
    
    self.img.layer.masksToBounds = YES;
    self.img.layer.cornerRadius = [UIScreen mainScreen].bounds.size.width * 0.8 / 2;
    
    self.lrcView.contentSize = CGSizeMake(self.view.bounds.size.width * 2, 0);
    self.lrcView.lrcLabel = self.lrcLabel;
    self.lrcView.delegate = self;
    self.lrcView.pagingEnabled = YES;
    
    [self.progress setThumbImage:[UIImage imageNamed:@"player_slider_playback_thumb"] forState:UIControlStateNormal];
}
#pragma mark 设置相关信息
- (void)setUIMessageWithModel:(MusicModel *)model
{
    self.MusicName.text = model.name;
    self.singer.text = model.singer;
    
    [self.img setImage:[UIImage imageNamed:model.icon]];
    
    self.lrcView.lrcName = model.lrcname;
}
#pragma mark 播放音乐
- (void)playMusic
{
    self.audioPlayer = [[AVAudioPlayer alloc]initWithContentsOfURL:[[NSBundle mainBundle]URLForResource:self.model.filename withExtension:nil] error:nil];
    self.audioPlayer.delegate = self;
    
    [self.audioPlayer prepareToPlay];
    
    if (!self.audioPlayer.isPlaying) {
        [self.audioPlayer play];
    }
//    获取音乐信息
    self.currentTime.text = [NSString stringWithTime:self.audioPlayer.currentTime];
    self.totalTIme.text = [NSString stringWithTime:self.audioPlayer.duration];
//    图片旋转动画
    [self.img.layer startAnimate];
    
    self.playBtn.selected = YES;
}
#pragma mark 按钮方法
- (IBAction)previousAndNextBtnAction:(UIButton *)sender {
    if (sender.tag == 11) {
        [self switchMusicWhetherNext:NO];
    }else
    {
        [self switchMusicWhetherNext:YES];
    }
}
- (IBAction)playBtnActiosn:(UIButton *)sender {
//    判断是否处于播放状态
    if (sender.isSelected) {
        [sender setSelected:NO];
        [self.img.layer pauseAnimate];
        [self.audioPlayer pause];
        [self.progressTimer pauseTimer];
    }else {
        [sender setSelected:YES];
        [self.img.layer resumeAnimate];
        [self.audioPlayer play];
        [self.progressTimer resumeTimer];
    }
}
#pragma mark 切换歌曲方法
- (void)switchMusicWhetherNext:(BOOL)flag
{
    if (flag) {
        if (self.currentIndex == self.data.count - 1) {
            self.currentIndex = 0;
        }else
        {
            self.currentIndex = self.currentIndex + 1;
        }
    }else
    {
        if (self.currentIndex == 0) {
            self.currentIndex = self.data.count - 1;
        }else
        {
            self.currentIndex = self.currentIndex - 1;
        }
    }
    self.model = self.data[self.currentIndex];
    
    if (self.audioPlayer) {
        [self.audioPlayer stop];
        self.audioPlayer = nil;
    }
    [self setUIMessageWithModel:self.model];
    
    [self playMusic];
    
}
#pragma mark 状态栏设成白色
- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}
#pragma mark 设置进度定时器
- (void)addProgressTimer
{
    self.progressTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(updateProgressInfo) userInfo:nil repeats:YES];
    [[NSRunLoop mainRunLoop] addTimer:self.progressTimer forMode:NSRunLoopCommonModes];
}
#pragma mark 更新时间进度
- (void)updateProgressInfo
{
    self.currentTime.text = [NSString stringWithTime:self.audioPlayer.currentTime];
    self.progress.value = self.audioPlayer.currentTime / self.audioPlayer.duration;
}
#pragma mark 设置歌词定时器
- (void)addLrcTimer
{
    self.lrcTimer = [CADisplayLink displayLinkWithTarget:self selector:@selector(updateLrc)];
    [self.lrcTimer addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
}
#pragma mark 更新歌词
- (void)updateLrc
{
    self.lrcView.currentTime = self.audioPlayer.currentTime;
}
#pragma mark 停止定时器
- (void)stopTimer:(int)flag
{
//flag: 0.停止进度定时器 1.停止歌词定时器 2.停止所有定时器
    switch (flag) {
        case 0:
            if (self.progressTimer) {
                [self.progressTimer invalidate];
                self.progressTimer = nil;
            }
            break;
        case 1:
            if (self.lrcTimer) {
                [self.lrcTimer invalidate];
                self.lrcTimer = nil;
            }
            break;
        case 2:
            if (self.progressTimer) {
                [self.progressTimer invalidate];
                self.progressTimer = nil;
            }
            if (self.lrcTimer) {
                [self.lrcTimer invalidate];
                self.lrcTimer = nil;
            }
            break;
        default:
            break;
    }
}
#pragma mark 滑动事件方法
- (IBAction)touchDown_StartChange:(UISlider *)sender {
    [self.progressTimer pauseTimer];
}
- (IBAction)touchDragInside_EndChange:(UISlider *)sender {
    self.audioPlayer.currentTime = sender.value * self.audioPlayer.duration;
    [self.progressTimer resumeTimer];
}
- (IBAction)valueChanged:(UISlider *)sender {
    NSTimeInterval currentTime = sender.value *self.audioPlayer.duration;
    self.currentTime.text = [NSString stringWithTime:currentTime];
}
- (IBAction)tapSlider:(UITapGestureRecognizer *)sender {
//    1.获取用户点击的点
    CGPoint point = [sender locationInView:sender.view];
//    2.点击的点在进度条中的比例
    CGFloat ratio = point.x / self.progress.bounds.size.width;
//    3.获取当前应播放的时间
    NSTimeInterval currentTime = self.audioPlayer.duration * ratio;
    self.audioPlayer.currentTime = currentTime;
//    4.更新进度信息
    [self updateProgressInfo];
}
#pragma mark - AVAudioPlayerDelegate
- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{
    if (flag) {
        [self switchMusicWhetherNext:YES];
    }
}
#pragma mark - UIScrollViewDelegate
- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
//    取出scrollView偏移量
    CGPoint Offset = scrollView.contentOffset;
    CGFloat offsetRatio = Offset.x / scrollView.bounds.size.width;
    
//    设置iconView和歌词Label的透明度
    self.img.alpha = 1 - offsetRatio;
    self.lrcLabel.alpha = 1 - offsetRatio;
}
@end

File: MusicDemoSwift/MusicDemoSwift/Controller/DetailViewController.swift
//
//  DetailViewController.swift
//  MusicDemoSwift
//
//  Created by 沈凯 on 2017/11/13.
//  Copyright © 2017年 Ssky. All rights reserved.
//

import UIKit
import AVFoundation
import MediaPlayer

class DetailViewController: UIViewController {
    
    var model: MusicModel!
    var data: [MusicModel]!
    var currentIndex: NSInteger!
//    歌名
    @IBOutlet weak var musicName: UILabel!
//    歌者
    @IBOutlet weak var singer: UILabel!
//    图片
    @IBOutlet weak var img: UIImageView!
//    进度
    @IBOutlet weak var progress: UISlider!
//    播放按钮
    @IBOutlet weak var playBtn: UIButton!
//    当前时间
    @IBOutlet weak var currentTime: UILabel!
//    总体时间
    @IBOutlet weak var totalTIme: UILabel!
//    歌词
    @IBOutlet weak var lrcLabel: LrcLabel!
//    歌词视图
    @IBOutlet weak var lrcView: LrcView!
//    播放器
    var audioPlayer: AVAudioPlayer!
//    进度的定时器
    var progressTimer: Timer!
//    歌词定时器
    var lrcTimer: CADisplayLink!
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        becomeFirstResponder()
        UIApplication.shared.beginReceivingRemoteControlEvents()
    }
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        resignFirstResponder()
        UIApplication.shared.endReceivingRemoteControlEvents()
    }
    override func viewDidLoad() {
        super.viewDidLoad()
//        UI设置
        setUI()
//        播放音乐
        playMusic()
//        后台播放
        playingBackground()
//        设置进度定时器
        addProgressTimer()
//        设置歌词定时器
        addLrcTimer()
//        状态栏设成白色
        UIApplication.shared.statusBarStyle = .lightContent
    }
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        stopTimer(2)
    }

//    MARK: UI设置
    func setUI() {
        setUIMessageWith(model)
        img.layer.masksToBounds = true
        img.layer.cornerRadius = CGFloat(UIScreen.main.bounds.size.width * 0.8 / 2)
        lrcView.contentSize = CGSize(width: view.bounds.size.width * 2, height: 0)
        lrcView.lrcLabel = lrcLabel
        lrcView.delegate = self
        lrcView.isPagingEnabled = true
        progress.setThumbImage(UIImage(named: "player_slider_playback_thumb"), for: .normal)
    }
//    MARK: 设置相关信息
    func setUIMessageWith(_ model: MusicModel) {
        musicName.text = model.name
        singer.text = model.singer
        img.image = UIImage.init(named: model.icon)
        lrcView.lrcName = model.lrcname
    }
//    MARK: 播放音乐
    func playMusic() {
        audioPlayer = try? AVAudioPlayer(contentsOf: Bundle.main.url(forResource: model.filename, withExtension: nil)!)
        audioPlayer.delegate = self
        audioPlayer.prepareToPlay()
        if !audioPlayer.isPlaying {
            audioPlayer.play()
        }
//        获取音乐信息
        currentTime.text = String.stringWithTime(time: audioPlayer.currentTime)
        totalTIme.text = String.stringWithTime(time: audioPlayer.duration)
//        图片旋转动画
        img.layer.startAnimate()
        playBtn.isSelected = true
//        设置锁屏信息显示
        setLockScreenDisplay()
        
    }
//    MARK: 设置锁屏信息显示
    func setLockScreenDisplay() {
        var info = Dictionary<String, Any>()
        info[MPMediaItemPropertyTitle] = model.name//歌名
        info[MPMediaItemPropertyArtist] = model.singer//作者
//        [info setObject:self.model.filename forKey:MPMediaItemPropertyAlbumTitle];//专辑名
        info[MPMediaItemPropertyAlbumArtist] = model.singer//专辑作者
        info[MPMediaItemPropertyArtwork] = MPMediaItemArtwork(image: UIImage.init(named: model.icon)!)//显示的图片
        info[MPMediaItemPropertyPlaybackDuration] = audioPlayer.duration//总时长
        info[MPNowPlayingInfoPropertyPlaybackRate] = 1.0//播放速率
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }
//    MARK: 后台播放
    func playingBackground() {
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(AVAudioSessionCategoryPlayback)
        try? session.setActive(true)
    }
//    MARK: 按钮方法
    @IBAction func previousAndNextBtnAction(_ sender: UIButton) {
        if sender.tag == 11 {
            switchMusicWhetherNext(false)
        }else {
            switchMusicWhetherNext(true)
        }
    }
    @IBAction func playBtnActiosn(_ sender: UIButton) {
//        判断是否处于播放状态
        if sender.isSelected {
            resumeAndPauseWhtherPause(true)
        }else {
            resumeAndPauseWhtherPause(false)
        }
    }
    //    MARK: 设置进度定时器
    func addProgressTimer() {
        progressTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(self.updateProgressInfo), userInfo: nil, repeats: true)
        RunLoop.main.add(progressTimer, forMode: .commonModes)
    }
//    MARK: 更新时间进度
    @objc func updateProgressInfo() {
        currentTime.text = String.stringWithTime(time: audioPlayer.currentTime)
        progress.value = Float(audioPlayer.currentTime / audioPlayer.duration)
    }
//    MARK: 设置歌词定时器
    func addLrcTimer() {
        lrcTimer = CADisplayLink(target: self, selector: #selector(self.updateLrc))
        lrcTimer.add(to: RunLoop.main, forMode: .commonModes)
    }
//    MARK: 更新歌词
    @objc func updateLrc() {
        lrcView.currentTime = audioPlayer.currentTime
    }
//    MARK: 停止定时器
    func stopTimer(_ flag: Int) {
//        flag: 0.停止进度定时器 1.停止歌词定时器 2.停止所有定时器
        switch flag {
        case 0:
            if (progressTimer != nil) {
                progressTimer.invalidate()
                progressTimer = nil
            }
        case 1:
            if (lrcTimer != nil) {
                lrcTimer.invalidate()
                lrcTimer = nil
            }
        case 2:
            if (progressTimer != nil) {
                progressTimer.invalidate()
                progressTimer = nil
            }
            if (lrcTimer != nil) {
                lrcTimer.invalidate()
                lrcTimer = nil
            }
        default:
            break
        }
    }
//    MARK: 播放暂停方法
    func resumeAndPauseWhtherPause(_ status: Bool) {
        playBtn.isSelected = !playBtn.isSelected
        var info = MPNowPlayingInfoCenter.default().nowPlayingInfo
        if status {
            img.layer.pauseAnimate()
            audioPlayer.pause()
            progressTimer.pauseTimer()
            info![MPNowPlayingInfoPropertyPlaybackRate] = 0.00001
            info![MPNowPlayingInfoPropertyElapsedPlaybackTime] = audioPlayer.currentTime
        }else {
            img.layer.resumeAnimate()
            audioPlayer.play()
            progressTimer.resumeTimer()
            info![MPNowPlayingInfoPropertyPlaybackRate] = 1.0
        }
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }
//    MARK: 切换歌曲方法
    func switchMusicWhetherNext(_ flag: Bool) {
        if flag {
            if currentIndex == data.count - 1 {
                currentIndex = 0
            }else {
                currentIndex = currentIndex + 1
            }
        }else {
            if currentIndex == 0 {
                currentIndex = data.count - 1
            }else {
                currentIndex = currentIndex - 1
            }
        }
        model = data[currentIndex]
        if (audioPlayer != nil) {
            audioPlayer.stop()
            audioPlayer = nil
        }
        setUIMessageWith(model)
        playMusic()
        
        img.layer.resumeAnimate()
        progressTimer.resumeTimer()
    }
//    MARK: 锁屏控制
    override func remoteControlReceived(with event: UIEvent?) {
        let type = event?.subtype
        switch type! {
        case .remoteControlPlay:
            //播放
            resumeAndPauseWhtherPause(false)
        case .remoteControlPause:
            //暂停
            resumeAndPauseWhtherPause(true)
        case .remoteControlStop: break
            //停止
        case .remoteControlTogglePlayPause: break
            //切换播放暂停（耳机线控）
        case .remoteControlNextTrack:
            //下一首
            switchMusicWhetherNext(true)
        case .remoteControlPreviousTrack:
            //上一首
            switchMusicWhetherNext(false)
        case .remoteControlBeginSeekingBackward: break
            //开始快退
        case .remoteControlEndSeekingBackward: break
            //结束快退
        case .remoteControlBeginSeekingForward: break
            //开始快进
        case .remoteControlEndSeekingForward: break
            //结束快进
        default:
            break
        }
    }
//    MARK: 滑动事件方法
    @IBAction func touchDown_StartChange(_ sender: UISlider) {
        progressTimer.pauseTimer()
    }
    @IBAction func touchDragInside_EndChange(_ sender: UISlider) {
        audioPlayer.currentTime = Double(sender.value) * audioPlayer.duration
        progressTimer.resumeTimer()
    }
    @IBAction func valueChanged(_ sender: UISlider) {
        let current = Double(sender.value) * audioPlayer.duration
        currentTime.text = String.stringWithTime(time: current)
    }
    @IBAction func tapSlider(_ sender: UITapGestureRecognizer) {
//        1.获取用户点击的点
        let point = sender.location(in: sender.view)
//        2.点击的点在进度条中的比例
        let ratio = point.x / progress.bounds.size.width
//        3.获取当前应播放的时间
        let currentTime = audioPlayer.duration * Double(ratio)
        audioPlayer.currentTime = currentTime
//        4.更新进度信息
        updateProgressInfo()
    }
}
//    MARK: - UIScrollViewDelegate
extension DetailViewController: UIScrollViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
//        取出scrollView偏移量
        let Offset: CGPoint = scrollView.contentOffset
        let offsetRatio: CGFloat = Offset.x / scrollView.bounds.size.width
//        设置iconView和歌词Label的透明度
        img.alpha = 1 - offsetRatio
        lrcLabel.alpha = 1 - offsetRatio
    }
}
//    MARK: - AVAudioPlayerDelegate
extension DetailViewController: AVAudioPlayerDelegate {

    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        if flag {
            switchMusicWhetherNext(true)
        }
    }
}
